primitives/src/vid/advz/namespace.rs

```diff
@@ -0,0 +1,355 @@
+// Copyright (c) 2023 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! Implementation of [`Namespacer`] for `Advz`.
+
+use ark_poly::EvaluationDomain;
+use jf_utils::{bytes_to_field, compile_time_checks};
+
+use super::{
+    AffineRepr, Debug, DenseUVPolynomial, Digest, DynDigest, GenericAdvz, MerkleTreeScheme,
+    PolynomialCommitmentScheme, PrimeField, UnivariatePCS, Vec, Write,
+};
+use crate::{
+    alloc::string::ToString,
+    vid::{namespace::Namespacer, vid, VidError},
+};
+use ark_std::format;
+
+impl<P, T, H, V> Namespacer for GenericAdvz<P, T, H, V>
+where
+    // TODO ugly trait bounds https://github.com/EspressoSystems/jellyfish/issues/253
+    P: UnivariatePCS<Point = <P as PolynomialCommitmentScheme>::Evaluation>,
+    P::Evaluation: PrimeField,
+    P::Polynomial: DenseUVPolynomial<P::Evaluation>,
+    P::Commitment: From<T> + AsRef<T>,
+    T: AffineRepr<ScalarField = P::Evaluation>,
+    H: Digest + DynDigest + Default + Clone + Write,
+    V: MerkleTreeScheme<Element = Vec<P::Evaluation>>,
+    V::MembershipProof: Sync + Debug,
+    V::Index: From<u64>,
+{
+    // TODO should be P::Proof, not Vec<P::Proof>
+    // https://github.com/EspressoSystems/jellyfish/issues/387
+    type DataProof = Vec<P::Proof>;
+
+    fn data_proof(
+        &self,
+        payload: &Self::Payload,
+        start: usize,
+        len: usize,
+    ) -> crate::vid::VidResult<Self::DataProof> {
+        // check args: `start`, `len` in bounds for `payload`
+        if start + len > payload.as_slice().len() {
+            return Err(VidError::Argument(format!(
+                "start {} + len {} out of bounds for payload {}",
+                start,
+                len,
+                payload.as_slice().len()
+            )));
+        }
+
+        // index conversion
+        let (start_elem, len_elem) = self.range_byte_to_elem(start, len);
+        let (start_namespace, len_namespace) = self.range_elem_to_poly(start_elem, len_elem);
+        let start_namespace_byte = self.index_poly_to_byte(start_namespace);
+
+        // check args:
+        // TODO TEMPORARY: forbid requests that span multiple polynomials
+        if len_namespace > 1 {
+            return Err(VidError::Argument(format!(
+                "request spans {} polynomials, expect 1",
+                len_namespace
+            )));
+        }
+
+        // grab the `start_namespace`th polynomial
```

Comment:
Optional TODO in the future PR: save this computation.

Reply:
#397